src/Core/StaticConfigTrait.php

```diff
@@ -249,23 +249,53 @@ public static function parseDsn($dsn)
             throw new InvalidArgumentException('Only strings can be passed to parseDsn');
         }
 
-        $pattern = '/^(?P<scheme>[\w\\\\]+):\/\/((?P<username>.*?)(:(?P<password>.*?))?@)?' .
-            '((?P<host>[^?#\/:@]+)(:(?P<port>\d+))?)?' .
-            '(?P<path>\/[^?#]*)?(\?(?P<query>[^#]*))?(#(?P<fragment>.*))?$/';
-        preg_match($pattern, $dsn, $parsed);
-
-        if (!$parsed) {
-            throw new InvalidArgumentException("The DSN string '{$dsn}' could not be parsed.");
+        $scheme = $noScheme = '';
+        if (preg_match('{^(?<scheme>[\w\\\]+)(:(//)?)(?P<noScheme>(.*))}', $dsn, $matches)) {
+            $scheme = $matches['scheme'];
+            $noScheme = $matches['noScheme'];
         }
-        foreach ($parsed as $k => $v) {
-            if (is_int($k)) {
-                unset($parsed[$k]);
+
+        $parsed = parse_url('file://' . $noScheme);
+        if ($parsed === false) {
+            // Failed to parse.
+            // Try to encode the userinfo if the DSN string contains '@' marks.
+            for ($p = 0; ($p = strpos($noScheme, '@', $p)) !== false; ++$p) { // @codingStandardsIgnoreLine
```

Comment:
I would have to apologize to you. In the first place, the fallback code might be unnecessary. Even if we use `preg_match()`, we could distinguish empty strings from non-specified, by capturing `@`, `:`, `/`, `?` and `#` followed by user, password/port, path, query and fragment. And, since the host part cannot be an empty string, if it is a empty string in the result, then it would mean that the host part is not specified.
So we could enhance the current code without reverting it. We could check whether those characters were captured or not before unsetting empty url components [here](https://github.com/cakephp/cakephp/blob/8a91fca6b147ebc0eb39c77ef330e357207dfd5a/src/Core/StaticConfigTrait.php#L264-L266). The big advantage of this approach is that we could return consistent results in both PHP5 and PHP7. Would you prefer this approach over the current approach in this pull request?